### PR TITLE
feat(rum-core): support regex same origin

### LINKED ISSUE #943: distributedTracingOrigins wildcard support
Hi, 

Would it be possible to add wildcard support, or regex pattern matching as an option to the tracing origins option?

I think it would be rather simple change here: https://github.com/elastic/apm-agent-rum-js/blob/master/packages/rum-core/src/common/utils.js#L120

Thank you,
Peter

### PR DESCRIPTION
**Pull request by baruchiro**
close #943 

I didn't test locally because of #972 and other things related to initialize dev env.

### CODE REVIEW

File: packages/rum-core/src/common/utils.js [hunk 1]
```diff
@@ -155,6 +155,8 @@ function checkSameOrigin(source, target) {
         isSame = checkSameOrigin(source, t)
       }
     })
+  } else if (typeof target === 'object' && target.constructor === RegExp) {
```
Comment (vigneshshanmugam) [inline packages/rum-core/src/common/utils.js:158]: you can check for `test === "function"` similar to https://github.com/elastic/apm-agent-rum-js/blob/8dda5c57a55f815f3252f7e2e769191efeefd7fe/packages/rum-core/src/performance-monitoring/transaction-service.js#L391

Comment (baruchiro) [reply packages/rum-core/src/common/utils.js]: Thanks, I think there are a lot of options, and each option assumes another thing.

But if we already have it, I will use it.

@vigneshshanmugam Solved

**Review by vigneshshanmugam — COMMENTED** [review]
Looks good, can you please fix the formatting and should be good to go.

**Comment by baruchiro** [comment]
@vigneshshanmugam sorry, rebased

**Comment by vigneshshanmugam** [comment]
@baruchiro Check this test error. 

```
[2021-03-07T19:16:39.752Z] node-puppeteer         | 	TypeError: Cannot read property 'test' of undefined

[2021-03-07T19:16:39.752Z] node-puppeteer         | 	    at <Jasmine>

[2021-03-07T19:16:39.752Z] node-puppeteer         | 	    at Module.checkSameOrigin (webpack:///src/common/utils.js:152:1 <- test/common/utils.spec.js:1561:28)

[2021-03-07T19:16:39.752Z] node-puppeteer         | 	    at UserContext.<anonymous> (webpack:///test/common/utils.spec.js:126:1 <- test/common/utils.spec.js:2087:77)

[2021-03-07T19:16:39.752Z] node-puppeteer         | 	    at <Jasmine>

[2021-03-07T19:16:39.752Z] node-puppeteer         | HeadlessChrome 90.0.4430 (Linux 0.0.0): Executed 97 of 274 (1 FAILED) (skipped 4) (0.859 secs / 0.769 secs)



```

File: packages/rum-core/src/common/utils.js [hunk 2]
```diff
@@ -149,6 +149,8 @@ function checkSameOrigin(source, target) {
   let isSame = false
   if (typeof target === 'string') {
     isSame = source === target
+  } else if (typeoftarget && typeof target.test === 'function') {
```
Comment (vigneshshanmugam) [inline packages/rum-core/src/common/utils.js:152]: `target && `

Comment (baruchiro) [reply packages/rum-core/src/common/utils.js:152]: Oh my godddddd I can't believe it!!
Sorry for your time 😞

Comment (baruchiro) [reply packages/rum-core/src/common/utils.js:152]: But I see you found the bug related to NPM7, now I hope I will get the git hooks and tests

**Review by vigneshshanmugam — APPROVED** [review]
Thanks @baruchiro Looks good. Do you want to do a follow up PR to update the docs about the new RegEx support ? 

Its here - https://github.com/elastic/apm-agent-rum-js/blob/master/docs/configuration.asciidoc#distributedtracingorigins

**Comment by baruchiro** [comment]
@vigneshshanmugam I'm not familiar enough with the source code, but working on the docs, I see there is a "hidden assumption" that we are comparing `URL.origin` with `URL.origin`.

I would like to write an integration test for declaring an expectation about `distributedTracingOrigins` as `origin` as string/regex, but I think it will be very complicated for me (but I need to try to launch the dev env again).

WDYT?

**Comment by baruchiro** [comment]
@vigneshshanmugam And about #977, why do it in separated PR? Do you want me to add the commit to here?

**Comment by vigneshshanmugam** [comment]
@baruchiro  You can definitely add it to this same PR, if you don't mind going back and forth on this PR. It would be better to have it here. Whichever works best.

File: docs/configuration.asciidoc [hunk 3]
```diff
@@ -214,14 +214,14 @@ Distributed tracing is enabled by default, however you can use this configuratio
 * *Type:* Array
 * *Default:* `[]`
 
-This option is an array of strings and determines which origins should be monitored as part of distributed tracing.
+This option is an array of strings or regexes and determines which origins should be monitored as part of distributed tracing.
```
Comment (vigneshshanmugam) [inline docs/configuration.asciidoc:217]:
```suggestion
This option can be set to an array containing one or more Strings or RegExp objects and determines which origins should be monitored as part of distributed tracing.
```

File: docs/configuration.asciidoc [hunk 4]
```diff
@@ -214,14 +214,14 @@ Distributed tracing is enabled by default, however you can use this configuratio
...
 More specificly this option is consulted when the agent is about to add distributed tracing http header (`traceparent`) to a request.
-Please note that each item in the array should be a valid url and at least include the origin part (other parts of the url are ignored).
+Please note that this list testing the origin part of the url (other parts of the url are ignored), each item in the array should be a valid url or a regex to test the origin part.
```
Comment (vigneshshanmugam) [inline docs/configuration.asciidoc:219]:
```suggestion
Please note that each item in the array should be a valid URL containing the origin (other parts of the url are ignored) or a RegExp object. If an item in the array is String, an exact match will be performed. If it's a RegExp object, its test function will be called with the request origin. 
```

Comment (vigneshshanmugam) [reply docs/configuration.asciidoc:219]: @bmorelli25 Can you have a look at the docs change? Thanks!

File: docs/configuration.asciidoc [hunk 5]
```diff
@@ -214,14 +214,14 @@ Distributed tracing is enabled by default, however you can use this configuratio
 * *Type:* Array
 * *Default:* `[]`
 
-This option is an array of strings and determines which origins should be monitored as part of distributed tracing.
+This option is an array of strings or regexes and determines which origins should be monitored as part of distributed tracing.
 More specificly this option is consulted when the agent is about to add distributed tracing http header (`traceparent`) to a request.
```
Comment (bmorelli25) [inline docs/configuration.asciidoc:218]: Might as well fix the spelling of `specifically` while we're here 🙃 

```suggestion
More specifically, this option is consulted when the agent is about to add the distributed tracing HTTP header (`traceparent`) to a request.
```

File: docs/configuration.asciidoc (hunk 4 above)
Comment (bmorelli25) [reply docs/configuration.asciidoc:219]: I've slightly edited your above recommendation for clarity.

**Comment by bmorelli25** [comment]
@elasticmachine, run elasticsearch-ci/docs

File: docs/configuration.asciidoc (hunk 4 above)
Comment (bmorelli25) [reply docs/configuration.asciidoc]: Screen Shot 2021-03-09 at 12 24 25 PM

Huh, that's really neat. GitHub recognizes a suggestion edit as a contribution.

**Review by bmorelli25 — APPROVED** [review]
Docs LGTM. Thanks for your contribution, @baruchiro–and a special thanks from me for updating the docs too 🥇

**Comment by baruchiro** [comment]
Documentation is part of the task!

Thank you all for your patience!
Hope to be in touch on more topics.

**Review by jahtalab — APPROVED** [review]
Thanks for your contribution @baruchiro , LGTM!

**Comment by jahtalab** [comment]
@baruchiro We probably will release a minor version next week.

**Pull request merged**